src/v/datalake/catalog_schema_manager.cc

```diff
@@ -168,49 +167,81 @@ simple_schema_manager::get_table_info(
     }
 
     iceberg::transaction txn(std::move(load_res.value()));
+    const auto* schema = txn.table().get_schema(txn.table().current_schema_id);
 
-    // Check schema compatibility
-    auto type_copy = desired_type.copy();
-    auto get_res = get_ids_from_table_meta(table_id, txn.table(), type_copy);
-    if (get_res.has_error()) {
-        co_return get_res.error();
-    }
+    auto writer_struct_type_with_ids = writer_struct_type.copy();
+    auto fill_res = iceberg::try_fill_field_ids(
+      schema->schema_struct, writer_struct_type_with_ids);
+    if (fill_res == iceberg::ids_filled::no) {
+        auto merged_schema_struct_type = schema->schema_struct.copy();
+        auto merge_res = iceberg::merge_struct_types(
+          writer_struct_type, merged_schema_struct_type);
+        if (merge_res.has_error()) {
+            vlog(
+              datalake_log.warn,
+              "Failed to merge record schema with table schema {}: {}",
+              table_id,
+              merge_res.error());
+            co_return errc::not_supported;
+        }
+
+        // Make a copy so that we still have the original struct in case we need
+        // to log it on the error path.
+        auto merged_schema_with_evo = merged_schema_struct_type.copy();
+        auto get_res = apply_evolution_rules(
+          table_id, txn.table(), merged_schema_with_evo);
+        if (get_res.has_error()) {
+            co_return get_res.error();
+        }
 
-    if (!get_res.value()) {
-        // The desired schema is backwards compatible with the current table
-        // schema. Add the schema to the table.
-        auto update_res = co_await txn.set_schema(
-          iceberg::schema{
-            .schema_struct = std::move(type_copy),
-            .identifier_field_ids = {},
-          });
-        if (update_res.has_error()) {
-            auto msg = fmt::format(
-              "Failed trying to apply schema update to table {}: {}",
+        if (get_res.value()) {
+            vlog(
+              datalake_log.error,
+              "Applying evolution rules on merged schema resulted in a no-op. "
+              "This means there is a bug in one of: filling field ids, "
+              "merging, or applying evolution rules. desired_type={}, "
+              "merged_schema={}, merged_schema_with_evo={}",
               table_id,
-              update_res.error());
-            switch (update_res.error()) {
-            case iceberg::action::errc::shutting_down:
-                vlog(datalake_log.debug, "{}", msg);
-                co_return errc::shutting_down;
-            case iceberg::action::errc::io_failed:
-            case iceberg::action::errc::unexpected_state:
-                vlog(datalake_log.warn, "{}", msg);
-                co_return errc::failed;
+              writer_struct_type,
+              merged_schema_struct_type,
+              merged_schema_with_evo);
+            co_return errc::failed;
```

Comment:
Does this end up blocking translation? If it's a no-op, what goes wrong if we return success instead?

Reply:
looking again, it's sort of unclear what we're expressing here and to whom. logging "it's a bug" smells a bit funny to me.

if this branch should be unreachable by construction, maybe `vassert(!get_res.value())` has the semantics we want?